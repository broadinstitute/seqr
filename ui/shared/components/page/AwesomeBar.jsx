import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { withRouter } from 'react-router-dom'
import { Search } from 'semantic-ui-react'

import { HttpRequestHelper } from 'shared/utils/httpRequestHelper'

const AwesomebarSearch = styled(Search)`
  width: 100%;

  .ui.icon.input {
    max-width: 100%;
    width: ${props => props.inputwidth || '100%'};
  }
  
  .results {
    min-width: ${props => props.inputwidth || '100%'};
    width: fit-content !important;
  }
`

class AwesomeBar extends React.Component
{
  static propTypes = {
    categories: PropTypes.array,
    newWindow: PropTypes.bool,
    placeholder: PropTypes.string,
    history: PropTypes.object,
    inputwidth: PropTypes.string,
  }

  static defaultProps = {
    categories: ['projects', 'families', 'individuals', 'genes'],
  }

  constructor(props) {
    super(props)

    this.state = {}

    this.httpRequestHelper = new HttpRequestHelper(
      '/api/awesomebar',
      this.handleHttpSuccess,
      this.handleHttpError,
    )
  }

  componentWillMount() {
    this.resetComponent()
  }

  render() {
    return <AwesomebarSearch
      category
      selectFirstResult
      inputwidth={this.props.inputwidth}
      loading={this.state.isLoading}
      onResultSelect={this.handleResultSelect}
      onSearchChange={this.handleSearchChange}
      results={this.state.results}
      value={this.state.value}
      minCharacters={1}
      placeholder={this.props.placeholder || 'Search project, family, gene name, etc.'}
    />
  }

  handleHttpSuccess = (response, urlParams) => {
    if (urlParams.q === this.state.value) {
      this.setState({ isLoading: false, results: response.matches })
    }
  }

  handleHttpError = (response) => {
    this.setState({ isLoading: false })
    console.error(response)
  }

  resetComponent = () => {
    this.setState({ isLoading: false, results: {}, value: '' })
  }

  handleSearchChange = (e, obj) => {
    this.setState({ isLoading: true, value: obj.value })
    this.httpRequestHelper.get({
      q: obj.value,
      categories: this.props.categories || '',
    })
  }

  handleResultSelect = (e, obj) => {
    e.preventDefault()
    this.setState({ value: obj.result.title })
    if (this.props.newWindow) {
      window.open(obj.result.href, '_blank')
    } else {
      this.props.history.push(obj.result.href)
    }
  }
}

export { AwesomeBar as AwesomeBarComponent }


export default withRouter(AwesomeBar)

