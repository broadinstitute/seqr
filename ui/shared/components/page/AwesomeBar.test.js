import React from 'react'
import toJson from 'enzyme-to-json'
import { shallow, configure } from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import { AwesomeBarComponent, AwesomeBarFormInput } from './AwesomeBar'

configure({ adapter: new Adapter() })

test('shallow-render without crashing', () => {
  const wrapper = shallow(<AwesomeBarComponent />)

  expect(toJson(wrapper)).toMatchSnapshot()
})

test('shallow-render AwesomeBarFormInput  without crashing', () => {
  shallow(<AwesomeBarFormInput />)
})
