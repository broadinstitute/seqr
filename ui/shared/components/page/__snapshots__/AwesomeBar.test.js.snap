// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallow-render without crashing 1`] = `
<AwesomeBar__AwesomebarSearch
  category={true}
  loading={false}
  minCharacters={1}
  onResultSelect={[Function]}
  onSearchChange={[Function]}
  placeholder="Search project, family, gene name, etc."
  results={Object {}}
  selectFirstResult={true}
  value=""
/>
`;
